Item #8853: Ontology, Form insert and update

Sigmonia opened the pull request:
#### Rationale
Item #8853: Ontology, Form insert and update

#### Related Pull Requests
* https://github.com/LabKey/ontology/pull/43

#### Changes
* Add concept picker component to show OntologyBrowser on insert/update forms
* Fix domain designer marking all OntologyLookupOptions dirty when a referenced column is changed
* Enabled Initial concept to be passed to OntologyBrowserModal

Sigmonia committed:
v2.33.0-fb-conceptdisplaycolumn.0

Sigmonia committed:
Ontology changes to support showing the Concept Picker for the Insert/update pages
- Fix issue where the OntologyLookupOptions would update for any ontology change in designer
- Support marking Label/Import fields as not shown for Insert/Update
- Update OntologyBrowserModal to show an initial concept
- Added ConceptModel to export
- Added ConceptPicker component to export

Sigmonia committed:
v2.33.0-fb-conceptdisplaycolumn.1

Sigmonia committed:
Fix navigate to active node for initial modal opening

Sigmonia committed:
Merge develop into branch

Sigmonia committed:
Provisional release notes

Sigmonia committed:
v2.33.1-fb-conceptdisplaycolumn.2

cnathe commented on packages/components/src/index.ts:
```diff
@@ -1096,6 +1098,8 @@ export {
     OntologyBrowserPanel,
     OntologyConceptOverviewPanel,
     OntologyBrowserFilterPanel,
+    ConceptPicker,
```
minor, but the other components that we expose from this set are all prefixed with "Ontology". Should we be consistent and call this the "OntologyConceptPicker"?

Sigmonia replied:
fixed.

cnathe commented on packages/components/src/internal/components/ontology/ConceptPicker.tsx:
```diff
@@ -0,0 +1,58 @@
+    fieldName: string;
+    fieldLabel: string;
+    fieldValue?: string;
+    onConceptSelection?: (concept: ConceptModel) => void
```
should this be a required property? looks like the usage below doesn't check for undefined...and I'm guessing the usage in the LK module is always settings this.

Sigmonia replied:
done

cnathe commented on packages/components/src/internal/components/ontology/ConceptPicker.tsx:
```diff
@@ -0,0 +1,58 @@
+            <div>
+                <a className='show-toggle' onClick={togglePicker}>{`Find ${fieldLabel} By Tree`}</a>
+                {showPicker && (
+                    <OntologyBrowserModal initOntologyId={ontologyId} onApply={onApplyConcept} onCancel={togglePicker} title={"Ontology Browser"} initConcept={concept} />
```
maybe this was in the mockups, but "Ontology Browser" seems like a bit of an odd title for the modal. I'd suggest matching the text from the link (i.e. "Find <fieldLabel> By Tree").

Sigmonia replied:
done.

cnathe commented on packages/components/src/internal/components/ontology/ConceptPicker.tsx:
```diff
@@ -0,0 +1,58 @@
+
+    useEffect(() => {
+        if (fieldValue) {
+            fetchConceptForCode(fieldValue)
```
Nice! This implementation for looking up the concept label worked out very cleanly.

cnathe commented on packages/components/src/internal/components/ontology/OntologyBrowserModal.tsx:
```diff
@@ -1,31 +1,30 @@
-import React, { FC, memo, useCallback, useState } from 'react';
+import React, { FC, memo, useCallback, useEffect, useState } from 'react';
```
`useEffect` was added here, but I don't see a usage for it below. can it be removed?

Sigmonia replied:
removed.

cnathe commented on packages/components/src/internal/components/ontology/OntologyBrowserModal.tsx:
```diff
@@ -1,31 +1,30 @@
 
 import { OntologyBrowserPanel } from './OntologyBrowserPanel';
 import { ConceptModel } from './models';
+import { fetchConceptForCode } from './actions';
```
`fetchConceptForCode ` was added here, but I don't see a usage for it below. can it be removed?

Sigmonia replied:
removed

cnathe commented on packages/components/src/internal/components/ontology/OntologyLookupOptions.tsx:
```diff
@@ -66,7 +69,17 @@ export class OntologyLookupOptions extends PureComponent<Props, State> {
     onFieldChange = (evt: any): void => {
-        this.onChange(evt.target.id, evt.target.value);
+        const val = evt.target.value;
+        let changes = List<IFieldChange>([{id: evt.target.id, value: evt.target.value }]);
```
```suggestion
        const val = evt.target.value;
        let changes = List<IFieldChange>([{id: evt.target.id, value: val }]);
```

Sigmonia replied:
fixed

cnathe commented on packages/components/src/internal/components/ontology/OntologyLookupOptions.tsx:
```diff
@@ -66,7 +69,17 @@ export class OntologyLookupOptions extends PureComponent<Props, State> {
+        const val = evt.target.value;
+        let changes = List<IFieldChange>([{id: evt.target.id, value: evt.target.value }]);
+
+        const valFieldIdx = this.props.domainFields.findIndex(df => df.name === val);
```
it looks like this `onFieldChange` is also called on change for the source ontology. In that case the `val` isn't a domain field name, so does that work as expected? I suppose it would result in a `-1` and skip the if statement below, right?

Sigmonia replied:
Added a new handler for the Ontology selector just to be safe.

cnathe commented on packages/components/src/internal/components/ontology/OntologyLookupOptions.tsx:
```diff
@@ -66,7 +69,17 @@ export class OntologyLookupOptions extends PureComponent<Props, State> {
+        if (valFieldIdx > 0) {
+            const valFieldInsertId = createFormInputId(DOMAIN_FIELD_SHOWNININSERTVIEW, this.props.domainIndex, valFieldIdx);
+            const valFieldUpdateId = createFormInputId(DOMAIN_FIELD_SHOWNINUPDATESVIEW, this.props.domainIndex, valFieldIdx);
+            changes = changes.push({id:valFieldInsertId, value:false}, {id:valFieldUpdateId, value: false});
```
Nice! I think this implementation works well.

cnathe replied:
Do you think we need to handle the case for the field name that was "unset" as the ontology helper column? For example, if I pick "field A" and then change it to "field B", do I care that "field A" still has these changes? I assume no...that would get pretty tricky to keep track of.

Sigmonia replied:
As you say this gets complicated very fast and I don't think trying to track it adds much value for the effort...
So my inclination is to see if we get feedback on it.

cnathe replied:
Sounds good. Thanks.

Sigmonia committed:
CR review feedback changes
- Clean up unused imports
- Adjust display column format to be consistent with other locations
- Change ConceptOverviewModel to be a tooltip
  - Add bsStyle prop to LabelHelpTip to help with styling

Sigmonia committed:
CR review feedback changes
- Rename ConceptOverviewModal to ConceptOverviewTooltip

Sigmonia committed:
v2.33.1-fb-conceptdisplaycolumn.3

Sigmonia committed:
rename ConceptPicker to match other components

Sigmonia committed:
Clean up some unused references
Make onConceptSelect required
Adjust concpet picker modal's title

Sigmonia committed:
v2.33.1-fb-conceptdisplaycolumn.4

cnathe commented on packages/components/src/internal/components/ontology/ConceptOverviewPanel.spec.tsx:
```diff
@@ -119,21 +119,21 @@ describe('ConceptOverviewModal', () => {
 
     test('no concept', () => {
-        const wrapper = mount(<ConceptOverviewModal concept={undefined} onClose={onCloseFn} />);
+        const wrapper = mount(<ConceptOverviewTooltip concept={undefined} onClose={onCloseFn} />);
```
Up at line 108, should change the `describe` as well to `ConceptOverviewTooltip`

cnathe commented on packages/components/src/internal/components/ontology/OntologyConceptAnnotation.tsx:
```diff
@@ -106,14 +102,17 @@ export const OntologyConceptAnnotationImpl: FC<OntologyConceptAnnotationImplProp
-                                        className="domain-validator-link domain-annotation-item"
-                                        onClick={toggleConceptModal}
+                                        className={classNames("domain-annotation-item", isFieldLocked ? "domain-text-label" : "domain-validator-link")}
+                                        onClick={isFieldLocked ? null : toggleSelectModal}
```
I pulled this up in the UI. It looks really good. I like the popover/hover details better and having the click on the current concept open the selector dialog. You don't have to do it as part of this PR, but now that the "initial concept" can be passed in to the selector, we could do that in this case.

Sigmonia committed:
merge develop into fb

Sigmonia committed:
v2.39.0-fb-conceptdisplaycolumn.5

Sigmonia committed:
merge develop into fb

Sigmonia committed:
Fix #42828: Show concept display label for concept annotations in DomainDesigners.

Sigmonia committed:
v2.41.0-fb-conceptdisplaycolumn.6

Sigmonia committed:
Fix snapshots due to addition of principalConceptDisplay field

Sigmonia committed:
Add ontology abbreviation to ConceptModel and Concept java interface serialization
Use ConceptModel's new ontology prop to set ConceptAnnotation's modal source ontology
fix jest tests
Adjust annotation concept display to default to principal code
Adjust OnotologyLookup domainproperty field change calculation to include field type check

Sigmonia committed:
merge develop into fb

Sigmonia committed:
v2.41.1-fb-conceptdisplaycolumn.7

Sigmonia committed:
Fix null reference

Sigmonia committed:
merge develop into fb

Sigmonia committed:
merge develop into fb

Sigmonia committed:
Fix up some of the new annotation tests and timings

Sigmonia committed:
merge develop into fb

Sigmonia committed:
v2.42.1-fb-conceptdisplaycolumn.9

Sigmonia committed:
Finalize release notes and version number
